pkg/deploy/apply_test.go

```diff
@@ -523,6 +524,124 @@ func TestWithAwaitableResource(t *testing.T) {
 	})
 }
 
+func TestWithDeletableResource(t *testing.T) {
+	resName := "res-name"
+	requireResourceExists := func(shouldExist bool) func(t *testing.T, clients *k8sClients, res resourceutil.Resource) {
+		return func(t *testing.T, clients *k8sClients, res resourceutil.Resource) {
+			gvr, err := resourceutil.FromGVKtoGVR(clients.discovery, *res.GroupVersionKind)
+			require.Nil(t, err)
+			_, err = clients.dynamic.Resource(gvr).
+				Get(context.TODO(), resName, metav1.GetOptions{})
+
+			if shouldExist {
+				require.Nil(t, err)
+			} else {
+				require.NotNil(t, err)
+				require.True(t, apierrors.IsNotFound(err))
+			}
+		}
+	}
+	testCases := []struct {
+		desc        string
+		annotations map[string]interface{}
+		setup       func(t *testing.T, clients *k8sClients, res resourceutil.Resource)
+		requireFn   func(t *testing.T, clients *k8sClients, res resourceutil.Resource)
+	}{
+		{
+			desc: "Deletes annotated resources before apply",
+			annotations: map[string]interface{}{
+				deleteBeforeApplyAnnotation: "true",
+			},
+			setup: func(t *testing.T, clients *k8sClients, res resourceutil.Resource) {
+				gvr, err := resourceutil.FromGVKtoGVR(clients.discovery, *res.GroupVersionKind)
+				require.Nil(t, err)
+				_, err = clients.dynamic.Resource(gvr).
+					Create(context.TODO(), &res.Object, metav1.CreateOptions{})
+				require.Nil(t, err)
+			},
+			requireFn: requireResourceExists(false),
+		}, {
+			desc:        "Ignores non annotated resources",
+			annotations: map[string]interface{}{},
+			setup: func(t *testing.T, clients *k8sClients, res resourceutil.Resource) {
+				gvr, err := resourceutil.FromGVKtoGVR(clients.discovery, *res.GroupVersionKind)
+				require.Nil(t, err)
+				_, err = clients.dynamic.Resource(gvr).
+					Create(context.TODO(), &res.Object, metav1.CreateOptions{})
+				require.Nil(t, err)
+			},
+			requireFn: requireResourceExists(true),
+		}, {
+			desc:        "Correctly handles non existing resources",
```

Comment:
Why non existing resources? Here the annotations should be set, otherwise this is a noop